File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<int> nodesBetweenCriticalPoints(ListNode* head) {
        vector<int> arr;

        ListNode* temp = head;

        while (temp){
            arr.push_back(temp->val);
            temp = temp->next;
        }

        int n = arr.size();
        int i = 1;
        vector<int> crit;

        while (i < n-1){
            if ((arr[i] < arr[i-1] && arr[i] < arr[i+1])|| (arr[i] > arr[i-1] && arr[i] > arr[i+1])){
                crit.push_back(i);
                i++;
            }else{
                i++;
            }
        }

        n = crit.size();
        if (n <= 1) return {-1 , -1};
        int mini = INT_MAX;
        for (int i = 1 ; i < n ; i++){
            mini = min(mini , crit[i] - crit[i-1]);
        }

        return {mini , (crit[n-1] - crit[0])};


    }
};

File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        arr = []

        temp = head

        while temp:
            arr.append(temp.val)
            temp = temp.next
        
        i = 1
        n = len(arr)
        crit = []
        while i < n-1:
            if (arr[i]<arr[i+1] and arr[i]<arr[i-1]) or (arr[i]>arr[i+1] and arr[i]>arr[i-1]):
                crit.append(i)
                i+=1
            else:
                i+=1
        
        ans = []

        if len(crit) <= 1:
            return [-1 , -1]
        
        mini = float('inf')
        for i in range(1 , len(crit)):
            mini = min(mini , crit[i] - crit[i-1])
        
        return [mini , crit[-1] - crit[0]]
